core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -543,11 +538,16 @@ private[spark] class Executor(
         // requires access to properties contained within (e.g. for access control).
         Executor.taskDeserializationProps.set(taskDescription.properties)
 
-        updateDependencies(
+        val updated = updateDependencies(
           taskDescription.artifacts.files,
           taskDescription.artifacts.jars,
           taskDescription.artifacts.archives,
           isolatedSession)
+        if (updated) {
+          // reset the thread class loader
+          val newIsolatedSession = getIsolatedSession(taskDescription)
+          Thread.currentThread.setContextClassLoader(newIsolatedSession.replClassLoader)
+        }
```

Comment:
`updateDependencies` is called in several places, could we integrate this logic into the method itself to prevent cases where we end up using the outdated instance? 

Maybe we can leave the `Thread.currentThread.setContextClassLoader(newIsolatedSession.replClassLoader)` part here